src/compiler/scala/tools/nsc/transform/patmat/Logic.scala

```diff
@@ -205,39 +247,50 @@ trait Logic extends Debugging  {
       def simplifyProp(p: Prop): Prop = p match {
         case And(fv)     =>
           // recurse for nested And (pulls all Ands up)
-          val ops = fv.map(simplifyProp) - True // ignore `True`
-
           // build up Set in order to remove duplicates
-          val opsFlattened = ops.flatMap {
-            case And(fv) => fv
-            case f       => Set(f)
-          }.toSeq
+          val opsFlattenedBuilder = collection.immutable.Set.newBuilder[Prop]
+          for (prop <- fv) {
+            val simplified = simplifyProp(prop)
+            if (simplified != True) { // ignore `True`
+              simplified match {
+                case And(fv) => fv.foreach(opsFlattenedBuilder += _)
+                case f => opsFlattenedBuilder += f
+              }
+            }
+          }
+          val opsFlattened = opsFlattenedBuilder.result()
 
-          if (hasImpureAtom(opsFlattened) || opsFlattened.contains(False)) {
+          if (opsFlattened.contains(False) || hasImpureAtom(opsFlattened)) {
             False
           } else {
-            opsFlattened match {
-              case Seq()  => True
-              case Seq(f) => f
-              case ops    => And(ops: _*)
+            opsFlattened.size match {
```

Comment:
Would (opsFlattened.size: @switch) help here?

Reply:
It would be useful documentation, but even without it the pattern matcher will compile to a lookup/table-switch if possible.

Reply:
@retronym Yeah, I use it like I use @tailrec ;)